### PR TITLE
Make models scriptable

### PR DESCRIPTION
**Pull request by frostedoyster**
The transition to equistore-torch was very smooth, although it seems to slow the code down.
Compilation of the alchemical example results in a cryptic error message

### CODE REVIEW

**Comment by frostedoyster** [comment]
For reference:

```
       50   11.598    0.232   11.598    0.232 {method 'run_backward' of 'torch._C._EngineBase' objects}
   505162    7.301    0.000    7.646    0.000 labels.py:618(position)
       25    3.615    0.145   17.116    0.685 spherical_expansions.py:96(forward)
       25    1.976    0.079    4.289    0.172 power_spectrum.py:14(forward)
       25    1.083    0.043    1.083    0.043 {method 'unbind' of 'torch._C._TensorBase' objects}
    15625    0.749    0.000    0.749    0.000 {built-in method torch.where}
       25    0.637    0.025    9.399    0.376 one_hot.py:6(one_hot)
   339500    0.350    0.000    0.899    0.000 inspect.py:90(ismethoddescriptor)
     9325    0.308    0.000    0.308    0.000 {built-in method torch._C._nn.linear}
   363750    0.300    0.000    1.604    0.000 inspect.py:285(isroutine)
 35050/25    0.299    0.000   26.095    1.044 module.py:1188(_call_impl)
    11948    0.250    0.000    0.250    0.000 {built-in method torch.tensor}
       25    0.247    0.010    0.783    0.031 adam.py:319(_single_tensor_adam)
  1039751    0.243    0.000    0.243    0.000 {built-in method builtins.hasattr}
```

**Comment by frostedoyster** [comment]
@Luthaf has this changed recently?

**Comment by Luthaf** [comment]
no, that's always been the case for equistore-torch. This looks like it is still using the Python version of equistore

**Comment by agoscinski** [comment]
It could be a convoluted installation with equistore.core and torch that causes errors which can easily happen. I updated the setup.py with most recent equistore-torch. Try to just `pip install .` with a fresh environment and run it again. My output in this case is

```
DESCRIPTION                                                                                                 
Random seed: 123123                              
Training on cpu                                                                                             
Reading dataset                                                                                                                                                                                                         
Shuffling and extracting from dataset                                                                       
Shuffling and extraction done                                                                               
normalize True                                                                                                                                                                                                          
61.00990338164251            
Average number of atoms per structure: 41.4                                                                 
All species: [21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 39, 40, 41, 42, 44, 45, 46, 47, 71, 72, 73, 74, 77, 78, 79]
Traceback (most recent call last):                                                                          
  File "/home/alexgo/code/torch_spex/examples/alchemical_model.py", line 260, in <module>
    model = Model(hypers, all_species, do_forces=do_forces).to(device)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^          
  File "/home/alexgo/code/torch_spex/examples/alchemical_model.py", line 110, in __init__                   
    self.spherical_expansion_calculator = SphericalExpansion(hypers, all_species, device=device)
                                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alexgo/code/torch_spex/torch_spex/spherical_expansions.py", line 88, in __init__              
    self.vector_expansion_calculator = VectorExpansion(hypers, self.all_species, device=device)
                                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alexgo/code/torch_spex/torch_spex/spherical_expansions.py", line 276, in __init__     
    self.radial_basis_calculator = RadialBasis(hypers_radial_basis, all_species, device=device)
                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/alexgo/code/torch_spex/torch_spex/radial_basis.py", line 31, in __init__
    self.all_species_labels = equistore.Labels(
                              ^^^^^^^^^^^^^^^^
AttributeError: module 'equistore' has no attribute 'Labels'
```
which is understandable.

**Comment by frostedoyster** [comment]
Thanks both! Indeed, there was an import I forgot to change. I now get the following error in `temporary_one_hot.py`:
```
'Any' object has no attribute or method 'reshape'.:
  File "/home/bigi/virtualenv-i/lib/python3.7/site-packages/torch_spex/temporary_one_hot.py", line 20

    indices = torch.where(
        original_labels.reshape(original_labels.size, 1) == possible_labels
        ~~~~~~~~~~~~~~~~~~~~~~~ <--- HERE
    )[1]
```

File: torch_spex/forces.py [hunk 1]
```diff
@@ -1,17 +1,21 @@
...
+        outputs=[energy],
+        inputs=[positions],
+        grad_outputs=grad_outputs,
         retain_graph=is_training,
         create_graph=is_training,
-    )
-    return [-single_structure_gradient for single_structure_gradient in gradient]
+    )[0]
+    if gradient is None:
+        raise ValueError("something went wrong")
```
Comment (agoscinski) [inline torch_spex/forces.py:16]: Might be more helpful for debugging
```suggestion
raise ValueError(
  "Unexpected None value for computed gradient. torch.autograd.grad input arguments:\n"
  f"  outputs={[energy]!r},\n"
  f"  inputs={[positions]!r},\n"
  f"  retain_graph={is_training},\n"
  f"  create_graph={is_training}")
```
This how the error message look like
image

File: torch_spex/radial_basis.py [hunk 2]
```diff
@@ -20,26 +20,32 @@ def __init__(self, hypers, all_species, device) -> None:
...
         elif hypers["type"] == "physical":
             self.n_max_l, self.spliner = get_physical_le_spliner(hypers["E_max"], hypers["r_cut"], hypers["scale"], hypers["normalize"], hypers["cost_trade_off"], device=device)
+        elif hypers["type"] == "custom":
+            # The custom keyword here allows the user to set splines from outside.
+            # After initialization of the model, the user will have to use generate_splines()
+            # from splines.py to generate splines and set these attributes (which we initialize to
+            # None for now) from outside
+            self.n_max_l, self.spliner = None, None
         else:
             raise ValueError("unsupported radial basis")
```
Comment (agoscinski) [inline torch_spex/radial_basis.py:30]:
```suggestion
            raise ValueError(f"unsupported radial basis type {hypers["type"]!r}")
```

File: torch_spex/radial_basis.py [hunk 3]
```diff
@@ -53,43 +59,45 @@ def __init__(self, hypers, all_species, device) -> None:
...
         radial_functions = self.spliner.compute(x)
 
         if self.is_alchemical:
-            one_hot_aj = torch.tensor(
-                equistore.one_hot(samples_metadata, self.all_species_labels),
-                dtype = torch.get_default_dtype(),
-                device = radial_functions.device
+            one_hot_aj = metatensor.torch.one_hot(
+                samples_metadata,
+                Labels(  # this labels object could be a class member, but its values need to be moved from CPU to GPU with the model
```
Comment (agoscinski) [inline torch_spex/radial_basis.py:73]: I don't know if you tried, but there is a `to` function in Labels to move to a different a device
https://lab-cosmo.github.io/metatensor/latest/reference/python/labels.html#metatensor.Labels.to

File: tests/test_spherical_expansions.py [hunk 4]
```diff
@@ -30,7 +29,7 @@ class TestEthanol1SphericalExpansion:
     batch = next(iter(loader))
 
     def test_vector_expansion_coeffs(self):
-        tm_ref = equistore.core.io.load_custom_array("tests/data/vector_expansion_coeffs-ethanol1_0-data.npz", equistore.core.io.create_torch_array)
+        tm_ref = metatensor.torch.load("tests/data/vector_expansion_coeffs-ethanol1_0-data.npz")
         # we need to sort both computed and reference pair expansion coeffs,
         # because ase.neighborlist can get different neighborlist order for some reasons
         tm_ref = sort_tm(tm_ref)
```
Comment (agoscinski) [inline tests/test_spherical_expansions.py:35]: Should be add the scripting test here? 

```python
vector_expansion = VectorExpansion(self.hypers, self.all_species, device=self.device)
vector_expansion = torch.jit.script(vector_expansion)
```

File: tests/test_spherical_expansions.py [hunk 5]
```diff
@@ -40,22 +39,22 @@ def test_vector_expansion_coeffs(self):
...
-        assert equistore.operations.allclose(tm_ref, tm, atol=1e-5, rtol=1e-5)
+        assert metatensor.torch.allclose(tm_ref, tm, atol=1e-5, rtol=1e-5)
 
     def test_spherical_expansion_coeffs_alchemical(self):
         with open("tests/data/expansion_coeffs-ethanol1_0-alchemical-hypers.json", "r") as f:
             hypers = json.load(f)
-        tm_ref = equistore.core.io.load_custom_array("tests/data/spherical_expansion_coeffs-ethanol1_0-alchemical-seed0-data.npz", equistore.core.io.create_torch_array)
+        tm_ref = metatensor.torch.load("tests/data/spherical_expansion_coeffs-ethanol1_0-alchemical-seed0-data.npz")
         torch.manual_seed(0)
         spherical_expansion_calculator = SphericalExpansion(hypers, self.all_species, device=self.device)
```
Comment (agoscinski) [inline tests/test_spherical_expansions.py:59]: Should be add the scripting test here? 

```suggestion
        spherical_expansion_calculator = SphericalExpansion(hypers, self.all_species, device=self.device)
        spherical_expansion_calculator = torch.jit.script(spherical_expansion_calculator)
```

File: tests/test_spherical_expansions.py [hunk 6]
```diff
@@ -90,26 +89,26 @@ class TestArtificialSphericalExpansion:
     batch = next(iter(loader))
 
     def test_vector_expansion_coeffs(self):
-        tm_ref = equistore.core.io.load_custom_array("tests/data/vector_expansion_coeffs-artificial-data.npz", equistore.core.io.create_torch_array)
+        tm_ref = metatensor.torch.load("tests/data/vector_expansion_coeffs-artificial-data.npz")
         tm_ref = sort_tm(tm_ref)
         vector_expansion = VectorExpansion(self.hypers, self.all_species, device=self.device)
```
Comment (agoscinski) [inline tests/test_spherical_expansions.py:94]:
```suggestion
        vector_expansion = VectorExpansion(self.hypers, self.all_species, device=self.device)
        vector_expansion = torch.jit.script(vector_expansion)
```

File: tests/test_spherical_expansions.py [hunk 7]
```diff
@@ -90,26 +89,26 @@ class TestArtificialSphericalExpansion:
...
         with torch.no_grad():
             tm = sort_tm(vector_expansion.forward(**self.batch))
-        assert equistore.operations.allclose(tm_ref, tm, atol=1e-5, rtol=1e-5)
+        assert metatensor.torch.allclose(tm_ref, tm, atol=1e-5, rtol=1e-5)
 
     def test_spherical_expansion_coeffs(self):
-        tm_ref = equistore.core.io.load_custom_array("tests/data/spherical_expansion_coeffs-artificial-data.npz", equistore.core.io.create_torch_array)
+        tm_ref = metatensor.torch.load("tests/data/spherical_expansion_coeffs-artificial-data.npz")
         spherical_expansion_calculator = SphericalExpansion(self.hypers, self.all_species, device=self.device)
```
Comment (agoscinski) [inline tests/test_spherical_expansions.py:101]:
```suggestion
        spherical_expansion_calculator = SphericalExpansion(self.hypers, self.all_species, device=self.device)
        spherical_expansion_calculator = torch.jit.script(spherical_expansion_calculator)
```

File: tests/test_spherical_expansions.py [hunk 8]
```diff
@@ -90,26 +89,26 @@ class TestArtificialSphericalExpansion:
...
         # I presume it is because we use 5 frames instead of just one
-        assert equistore.operations.allclose(tm_ref, tm, atol=3e-5, rtol=1e-5)
+        assert metatensor.torch.allclose(tm_ref, tm, atol=3e-5, rtol=1e-5)
 
     def test_spherical_expansion_coeffs_artificial(self):
         with open("tests/data/expansion_coeffs-artificial-alchemical-hypers.json", "r") as f:
             hypers = json.load(f)
-        tm_ref = equistore.core.io.load_custom_array("tests/data/spherical_expansion_coeffs-artificial-alchemical-seed0-data.npz", equistore.core.io.create_torch_array)
+        tm_ref = metatensor.torch.load("tests/data/spherical_expansion_coeffs-artificial-alchemical-seed0-data.npz")
         spherical_expansion_calculator = SphericalExpansion(hypers, self.all_species, device=self.device)
```
Comment (agoscinski) [inline tests/test_spherical_expansions.py:112]:
```suggestion
        spherical_expansion_calculator = SphericalExpansion(hypers, self.all_species, device=self.device)
        spherical_expansion_calculator = torch.jit.script(spherical_expansion_calculator)
```

**Review by agoscinski — COMMENTED** [review]
Looks good. Some radial basis changes seem to me intermingled here with the scriptability, but its okay. Main point is to add tests for torch scriptability, see individual comments.

File: torch_spex/radial_basis.py (hunk 3 above)
Comment (frostedoyster) [reply torch_spex/radial_basis.py:73]: Very good, I'll try if it works

**Review by agoscinski — APPROVED** [review]
I am okay with the PR, would try to merge the torch.jit.script tests in one test file to save some computation time, by not recomputing the neighborlist, but I don't know if this really saves much time.

**Pull request merged**